crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{
+    domain::competition::order,
+    tests::{
+        self,
+        cases::ToWei,
+        setup::{ab_order, ab_order_quote, ab_pmm_pool, ab_solution, ExpectedOrder, FeePolicy},
+    },
+};
+
+#[tokio::test]
+#[ignore]
+async fn surplus_protocol_fee_buy_order_not_capped() {
+    let side = order::Side::Buy;
+    let fee_policy = FeePolicy::Surplus {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+    };
+    let test_name = format!("Protocol Fee: {side:?} {fee_policy:?}");
+    let quote = ab_order_quote();
+    let pool = ab_pmm_pool(&quote);
+    let order = ab_order()
+        .kind(order::Kind::Limit)
+        .side(side)
+        .solver_fee(Some(10u32.to_wei()))
+        .fee_policy(fee_policy)
+        .quote(quote);
+    let test = tests::setup()
+        .name(test_name)
+        .pool(pool)
+        .order(order)
+        .solution(ab_solution())
+        .done()
+        .await;
+    let expected = ExpectedOrder {
+        name: ab_order().name,
+        executed_sell_amount: 100u32.to_wei(),
+        executed_buy_amount: 40u32.to_wei(),
```

Comment:
Can we make it so that from reading the test it's easy to understand the math? Ie, here I would expect us to set the limit price and executed price explicitly (which then makes it clear why we expect the following executed amounts).

Reply:
nit: In fact you can probably create a macro or TestCase struct which contains all fields (order type, limit, quote, executed amounts and expectations).

Reply:
Added explicit values and extracted all the common parts into a single function.